Rclean blog post (#646)
ropensci/roweb2: content/blog/2020-03-13-rclean/index.Rmd

---

> slug: "2020-03-13"

Reviewer: ```suggestion
slug: "rclean"
```

title: Automatic Code Cleaning in R with Rclean
# Delete the package_version line below if your post is not about a package
package_version: 1.1.8
author:
  - M.K. Lau
# Set the date below to the publication date of your post
date: 2020-03-13
# Minimal tags for a post about a community-contributed package 
# that has passed software peer review are listed below
# Consult the Technical Guidelines for information on choosing tags
tags:
  - Software Peer Review
  - packages
  - R
  - community
  - reproducibility
  - transparency

>   - clean code

Reviewer 2: Don't forget your package! :smile: 
```suggestion
  - clean code
  - Rclean
```

# the summary below will be used by e.g. Twitter cards
description: "Automatic isolation of code for reproducing results in R."
# If you have no preferred image for Twitter cards,
# delete the twitterImg line below 
# Note there is no '/' symbol before 'img' here
# if needed replace blog with technotes
# twitterImg: blog/2019/06/04/post-template/name-of-image.png
# 'output' is necessary to obtain index.md
# Do not commit index.html
output: 
  html_document:
    keep_md: true
bibliography: refs.bib
---

> Leave the code cleaner than you found it. 

-- R.C. Martin in *Clean Code*

The R language has become very popular among scientists and analysts
because it enables the rapid development of software and empowers
scientific investigation. However, regardless of the language used,
data analysis is usually complicated. Because of various project
complexities and time constraints, analytical software often reflects
these challenges. "What did I measure? What analyses are relevant to

> them? Do I need to transform the data? What's the function for the

Author: Good point. Now clarified with commit 4ffd414a.

> them? Do I need to transform the data? What's the function for the

Author: Good point. Fixed this with 4ffd414a.

is writing an R script, the ultimate goal is analytical results for
inference, such as a set of statistical analyses and associated
figures and tables. As the investigative process is inherently
iterative, this set of results is nearly always a subset of a much
larger set of possible ways to explore a dataset. There are many
statistical tests and visualizations and other representations that
can be employed in a myriad of ways depending on how the data are
processed. This commonly leads to lengthy, complicated scripts from
which researchers manually subset results, but likely never to be
refactored because of the difficulty in disentangling the code.

The [Rclean](https://docs.ropensci.org/rclean) package uses a
technique based on data provenance and network algorithms to isolate
code for a desired result *automatically*. The intent is to ease
refactoring for scientists that use R but do not have formal training
in software design and specifically with the "art" of creating clean
code, which in part is the development of an intuitive sense of how
code is and/or should be organized. However, manually culling code is

> tedious and potentially leads to errors, and we see

Author: This suggestion didn't quite capture the idea, so I've edited the text. See commit df7c9711.

> tedious and potentially leads to errors, and we see

Author: The suggested edit didn't quite capture the idea I was going for. I've added new text with df7c9711.

[Rclean](https://docs.ropensci.org/rclean) as a useful tool for
everyone at all levels of coding ability in R. Here, we'll cover
details on the implementation and design of the package, a general
example of how it can be used and thoughts on its future development.

> # How to Use Rclean to Write Cleaner Code

Author: Fixed a62b33fe.

## Installation

Through the helpful feedback from the ROpenSci community, the package

> has recently passed software review and a supporting article was

Reviewer 2: This would be a good place to link to the review issue and acknowledge the reviewers and editor.

Author: Great, happy to. Added a link here and also explicit acknowledgments in the concluding paragraph: de68b190.

fit.xx <- lm(A~B, data = x)
summary(fit.xx)
shapiro.test(residuals(fit.xx))

## Ah, it looks like I should probably transform A.
## Let's try that.
fit_sqrt_A <- lm(I(sqrt(A))~B, data = x)
summary(fit_sqrt_A)
shapiro.test(residuals(fit_sqrt_A))

## Looks good!

## After that. I came back and ran another analysis with 
## x2 and a new variable.

z <- c(rep("A", nrow(x2) / 2), rep("B", nrow(x2) / 2))
fit_anova <- aov(x2 ~ z, data = data.frame(x2 = x2[, 1], z))
summary(fit_anova)

```

So, let's say we've come to our script wanting to extract the code to
produce one of the results `fit_sqrt_A`, which is an analysis that is
the fitted model object for an analysis. We might want to double check
the results, and we also might need to use the code again for another
purpose, such as creating a plot of the patterns supported by the
test. Manually tracing through our code for all the variables used in
the test and finding all of the lines that were used to prepare them
for the analysis would be annoying and difficult, especially given the
fact that we have used "x" as a prefix for multiple unrelated objects
in the script. However, [Rclean](https://docs.ropensci.org/rclean) can

> do this easily via the `clean` function.

Reviewer 2: ```suggestion
do this easily via the `clean()` function.
```

```{R long}
library(Rclean)
clean(script, "fit_sqrt_A")

> ```

Author: Good point. I've updated the text to make this more transparent: 5b74da3b.

As you can see, [Rclean](https://docs.ropensci.org/rclean) has picked
through the tangled bits of code and found the minimal set of lines
relevant to our object of interest. This code can now be visually
inspected to adapt the original code or ported to a new, "refactored"

> script using `keep`.

Reviewer 2: ```suggestion
script using `keep()`.
```

```{R ex-keep, eval = FALSE}
fitSA <- clean(script, "fit_sqrt_A")
keep(fitSA)

```

This will pass the code to the clipboard for pasting into another
document. To write directly to a new file, a file path can be
specified.

```{R ex-keep-write, eval = FALSE}
fitSA <- clean(script, "fit_sqrt_A")
keep(fitSA, file = "fit_SA.R")

```

> To explore more possible variables to extract, the `get_vars` function

Reviewer 2: ```suggestion
To explore more possible variables to extract, the `get_vars()` function
```

can be used to produce a list of the variables (aka. objects) that are
created in the script.

```{R ex-get_vars-hid, eval = TRUE, echo = TRUE}

get_vars(script)

```

Especially when the code for different variables are entangled, it can
be useful to visual the code in order to devise an approach to

> cleaning. The `code_graph` function can also give us a visual of the

Reviewer 2: ```suggestion
cleaning. The `code_graph()` function can also give us a visual of the
```

code and the objects that they produce.

> ```{R ex-code_graph, dpi = 300, fig.cap = "Example of the plot produced by the code_graph function showing which lines of code produce which variables and which variables are used by other lines of code."}

Author: OK, I think I did this correctly. See a429c80b.

code_graph(script)

```

After examining the output from `get_vars` and `code_graph`, it is
possible that more than one object needs to be isolated. To do this is

> simple in `keep` by passing the set of desired objects to the *vars*

Reviewer 2: ```suggestion
After examining the output from `get_vars()` and `code_graph()`, it is
possible that more than one object needs to be isolated. To do this is
simple in `keep()` by passing the set of desired objects to the *vars*
```

argument.

```{R clean-mlti}

clean(script, vars = c("fit_sqrt_A", "fit_anova"))

```

Currently, libraries can not be isolated directly during the cleaning
process. So, the `get_libs` function provides a way to detect the
libraries for a given script. We just need to supply a file path and

> `get_libs` will return the libraries that are called by that script.

Reviewer 2: ```suggestion
process. So, the `get_libs()` function provides a way to detect the
libraries for a given script. We just need to supply a file path and
`get_libs()` will return the libraries that are called by that script.
```

```{R ex-get_libs}

get_libs(script)

```

# The Provenance Engine Under the Hood

> The `clean` function provides an effective way to remove code that is

Reviewer 2: ```suggestion
The `clean()` function provides an effective way to remove code that is
```

instead rely on file backups and informal versioning. Although we
can't give researchers more hours in the day, providing an easier and
more reliable means to remove unused code will lower the barrier to
creating better, cleaner code. Combined with the increasing use of
version control systems and digital notebooks, the practice of
"saving" analytical ideas in a script will become less common and code
quality will increase.

The process that [Rclean](https://docs.ropensci.org/rclean) uses
relies on the generation of data provenance.  The term provenance
means information about the origins of some object. Data provenance is
a formal representation of the execution of a computational process
(https://www.w3.org/TR/prov-dm/), to rigorously determine the the
unique computational pathway from inputs to results [@Carata2014]. To
avoid confusion, note that "data" in this context is used in a broad
sense to include all of the information generated during computation,
not just the data that are collected in a research project that are
used as input to an analysis. Having the formalized, mathematically
rigorous representation that data provenance provides guarantees that

> the analyses that [Rclean](https://docs.ropensci.org/rclean) conducts

Reviewer 2: ```suggestion
analyses conducted by [Rclean](https://docs.ropensci.org/rclean) 
```

are theoretically sound. Most importantly, because the relationships
defined by the provenance can be represented as a graph, it is
possible to apply network search algorithms to determine the minimum
and sufficient code needed to generate the chosen result in the

> `clean` function.

Reviewer 2: ```suggestion
`clean()` function.
```

There are multiple approaches to collecting data provenance, but
[Rclean](https://docs.ropensci.org/rclean) uses "prospective"

> provenance, which analyzes code and uses language specific information

Reviewer 2: ```suggestion
provenance, which analyzes code and uses language-specific information
```

to predict the relationship among processes and data
objects. [Rclean](https://docs.ropensci.org/rclean) relies on a

> library called CodeDepends to gather the prospective provenance for

Reviewer 2: ```suggestion
objects. [Rclean](https://docs.ropensci.org/rclean) relies on an
R package called CodeDepends to gather the prospective provenance for
```

each script. For more information on the mechanics of the CodeDepends
package, see [@R-CodeDepends]. To get an idea of what data provenance

> is, we can use `code_graph` function to get a graphical representation

Reviewer 2: ```suggestion
is, we can use the `code_graph()` function to get a graphical representation
```

of the prospective provenance generated for

> [Rclean](https://docs.ropensci.org/rclean) \ref{fig:prov-graph}.

Reviewer 2: As the figure is right below the text, I don't think a reference is necessary
```suggestion
[Rclean](https://docs.ropensci.org/rclean).
```

```{R load-script, echo = FALSE, results = "hide"}
library(CodeDepends)
library(Rclean)
```

> ```{R prov-graph, dpi = 300, fig.cap = "Network diagram of the prospective data provenance generated for an example script. Arrows indicate which lines of code (numbered) produced which objects (named).", echo = FALSE}

Reviewer 2: Two comments here:
1) If the numbered lines produce (->) named objects, what does it mean when an object leads (->) to a numbered line? e.g., 1 -> mat -> 2
2) See comments above on using the `hugoopts` chuck option to get the caption and alt text to work properly.

Author: I added more text here to explain this: b699bda2.

script <- system.file(
    "example", 
    "simple_script.R", 
    package = "Rclean")
code_graph(script)

```

All of this work with the provenance is to get the network
representation of relationships among functions and objects. The
provenance network is very powerful because we can now apply

> algorithms analyze the R script with respect to our results. This is

Reviewer 2: ```suggestion
algorithms to analyze the R script with respect to our results. This is
```

> what empowers the `clean` function, which takes the provenance and

Reviewer 2: ```suggestion
what empowers the `clean()` function, which takes the provenance and
```

> applies a a network search algorithm to determine the pathways leading

Reviewer 2: ```suggestion
applies a network search algorithm to determine the pathways leading
```

from inputs to outputs. In the process any objects or lines of code
that do not fall along that pathway are by definition not necessary to
produce the desired set of results and can therefore be removed. As
demonstrated in the example, this property of the provenance network
is what facilitates the robust isolation of the minimal code necessary
to generate the output we want.

One important topic to discuss is that
[Rclean](https://docs.ropensci.org/rclean) *does not* keep comments
present in code. This is the result of a limitation of the data
provenance collection, which currently does not assign them a
relationship in the provenance network. This is a general issue with
detecting the relationships between comments and code. For example,
comments at the end of lines are typically relevant to the line they
are on, but this is not a requirement and could refer to other
lines. Also, comments occupying their own lines usually refer to the
following lines, but this is also not necessarily the case either, as
there is no "linguistic" requirement for this. Although
[Rclean](https://docs.ropensci.org/rclean) cannot operate

> automatically on comments; however, comments in the original code

Reviewer 2: ```suggestion
automatically on comments, comments in the original code
```

remain untouched and can be used to inform the reduced code. Also, as

> the `clean` function is oriented toward isolating code based on a

Reviewer 2: ```suggestion
the `clean()` function is oriented toward isolating code based on a
```

specific result, the resulting code tends to naturally support the
generation of new comments that are higher level (e.g. "The following
produces a plot of the mean response of each treatment group."), and
lower level comments are not necessary because the code is simpler and

> clearer. This process of comments is an important part of writing

Reviewer 2: ```suggestion
clearer. This process of commenting is an important part of writing
```

better code. Lastly, although comments can serve an important role in
coding, it is worth reflecting on the statement in R.C. Martin's book
_Clean Code: A Handbook of Agile Software Craftsmanship_ where he
writes that, "Comments do not compensate for bad code."

# Concluding Remarks and Future Work

[Rclean](https://docs.ropensci.org/rclean) provides a simple, easy to
use tool for scientists who would like help refactoring code. Using
[Rclean](https://docs.ropensci.org/rclean), the code necessary to
produce a specified result (e.g., an object stored in memory or a
table or figure written to disk) can be easily and *reliably* isolated
even when tangled with code for other results. Tools, such as this,
that make it easier to produce transparent, accessible code will be an
important aid for improving scientific reproducibility
[@Pasquier2017].

Although the current implementation of
[Rclean](https://docs.ropensci.org/rclean) for minimizing code is
useful on its own, we see promise in connecting it with other
reproducibility tools. One example is the

> [reprex](https://reprex.tidyerse.org) package, which provides a simple

Reviewer 2: ```suggestion
[reprex](https://reprex.tidyverse.org) package, which provides a simple
```

processing comments (as discussed above), parsing control statements
and replicating random processes. Using retrospective provenance comes
at a cost, however. In order to gather it, the script needs to be
executed. When scripts are computationally intensive or contain bugs
that stop execution, then retrospective provenance can not be obtained
for part or all of the code. Some work has already been done in the
direction of implementing retrospective provenance for code cleaning
in R (see http://end-to-end-provenance.github.io). Although such costs
may present challenges, combining prospective and retrospective
provenance methods could provide a powerful and flexible solution.

We hope that [Rclean](https://docs.ropensci.org/rclean) makes writing
scientific software easier for the R community. We look forward to the
future progress of the package and other "code cleaning" tools. As an
open-source project, we are very receptive to feedback and help with
extending it. We invite people to use the package and get involved by
reporting bugs and suggesting or (hopefully) contributing
features. For more information please visit the project page on

> Github.

Reviewer 2: ```suggestion
[GitHub](https://github.com/ropensci/Rclean).
```
